### ISSUE TITLE
Proposal: Introduce a Relational Comparison Type Relationship

### ISSUE THREAD

**Issue by DanielRosenwasser** [Microsoft/TypeScript#10120]
# Problem

Before literal types and union types, it made sense that relational comparison operators and equality comparison operators had the same rules.
## Overly-permissive

However, with union types, relational comparisons don't always make sense.

``` ts
declare let x: string;
declare let y: string | number;

if (x < y) {
    // ...
}
```

Here, it's not clear that you'll get the "correct" behavior if `x` and `y` are not both `string`s, but TypeScript permits this. Issue #5156 covers this in more detail.
## Overly-strict

While the current "comparable" relationship is the most lax relationship, it is also too strict. For instance, consider the following example from #10119:

``` ts
function f(onethree: 1 | 3, two: 2) {
  return onethree < two;
}
```

Under the comparable relationship, TypeScript issues an error because the type `2` is not comparable with `1` nor is it comparable with `3`.
# Proposal

I propose a new _relational comparison_ relationship, in which for any types _S_ and _T_:
- If _T_ is a union type, then _S_ is only relationally comparable if _S_ is relationally comparable with each constituent of _T_.
- Otherwise, if _S_ is a union type, then _S_ is only relationally comparable to _T_ if each constituent of _S_ is relationally comparable to _T_.
- Otherwise, if _S_ is string-like, then _S_ is relationally comparable if _T_ is string-like or _Any_.
- Otherwise, if _S_ is number-like, then _S_ is relationally comparable if _T_ is number-like or _Any_.
- Otherwise, if _S_ is _Any_, then _S_ is relationally comparable if _T_ is number-like, string-like, or _Any_.

To note:
- _Boolean_ and other unmentioned types are never relationally comparable to anything. The rationale is that you probably didn't want to compare a _Boolean_ with a less-than operator.
- _Any_ is not relationally comparable to everything - for instance, you can't compare a _Boolean_ with an _Any_.
- There should not be a case where a mix of string-like and number-like types are ever relationally comparable.
- Type parameters are currently not covered here (see below).
# Open Questions

It's not clear how this behaves with type parameters. For instance, you might propose that we just check the type parameter's constraint here, but that's not enough. If you constrain `T` to `number | string`, then `T` will not be relationally comparable to itself because two distinct values of type `T` could each have different types at runtime. See https://github.com/Microsoft/TypeScript/issues/5156#issuecomment-231660972 for a concrete example.

**Comment by RyanCavanaugh** [comment]
> Otherwise, if S is number-like, then S is relationally comparable if T is number-like or Any.

One caveat is that it's probably not desirable to allow relational operations between differing enum types

**Comment by sandersn** [comment]
After some discussion, we decided in the 9/23 design meeting to just raise literal types to their base type in a comparison. This fixes #10119 and friends but not #5156.

**Comment by sandersn** [comment]
Fix is up at #12158
